src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -4164,11 +4164,128 @@ LogicalResult ONNXHardSwishOp::inferShapes(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ONNXIfOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+bool areIfTypesMergeable(Type ifResultType, Type branchResultType) {
```

Comment:
Can this merge function be used for all the joint of control flow, such as Loop? Can you attach a link to MLIR document that specifies the rule?

Reply:
I based the implementation on the MLIR documentation of the If op [here](https://github.com/onnx/onnx/blob/main/docs/Operators.md#If) where it says:

> "The return values in the `then_branch` and `else_branch` must be of the same data type. The `then_branch` and `else_branch` may produce tensors with the same element type and different shapes. If corresponding outputs from the then-branch and the else-branch have static shapes S1 and S2, then the shape of the corresponding output variable of the if-node (if present) must be compatible with both S1 and S2 as it represents the union of both possible shapes."

After re-reading this I notice they use the terminology "compatible" and "union", so I renamed `areIfTypesMergeable` to `areCompatibleIfTypes` and renamed `mergeIfTypes` to `unionOfIfTypes`.

I can try to reuse the logic for Loop. If it works, we can rename the methods to something more general and move them to Support/TypeUtilities. How about we leave that for a follow-up PR?